guidance/models/_azure_guidance.py

```diff
@@ -0,0 +1,167 @@
+import requests
+import os
+import base64
+import json
+import urllib.parse
+from ._model import Engine, Model, EngineCallResponse
+from .._chat import Phi3ChatTemplate
+from ._byte_tokenizer import ByteTokenizer
+
+
+class AzureGuidanceEngine(Engine):
+    """This connects to a remote guidance server on Azure and runs all computation using the remote engine."""
+
+    def __init__(self, server_url, max_streaming_tokens=1000, chat_template=None):
+        if (
+            server_url is None
+            or isinstance(server_url, str)
+            and len(server_url.strip()) == 0
+        ):
+            server_url = os.getenv("AZURE_GUIDANCE_URL", "")
+        elif not isinstance(server_url, str):
+            raise ValueError("server_url must contain a URL string.")
+
+        if not server_url.startswith("http"):
+            raise ValueError(
+                "AzureGuidance requires a remote model URL that starts with http"
+            )
+        self.server_url = server_url
+        self.max_streaming_tokens = max_streaming_tokens
+
+        if chat_template is None:
+            # TODO [PK]: obtain this from the server
```

Comment:
Do we need to know what the internal tokenisation of the model is? On the client side, don't we just take the `messages`, pack them into a byte string, run the grammar, and then format the result back into a list of messages? Nothing ever sees the model's tokenisation. There is token healing, but is that handled server or client side?

Reply:
We don't need to know the model's internal tokenization, but we do need to know the model's role tags.  The message format that most chat models are currently using isn't the best format for guidance because we might want to generate text within the assistant blocks without closing the block, or alternatively generate entire assistant/user/assistant/user/assistant turns using guidance grammars, which is easier to do using text based role tags.  To handle these scenarios, the API we're defining passes the chat role tags to the server as text without using lists of messages.  Token healing in this API is handled on the server.